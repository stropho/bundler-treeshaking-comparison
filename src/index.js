export * from './rollup'
export * from './terser'
export * from './webpack'
