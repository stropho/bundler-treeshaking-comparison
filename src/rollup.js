import { rollup } from 'rollup'

const ROLLUP_OUTPUT_OPTIONS = {
  file: 'output.js',
  format: 'commonjs',
}
const ROLLUP_INPUT_OPTIONS = {
}

const runRollup = async (inputOptions, outpuOptions) => {
  const bundle = await rollup(inputOptions)
  const result = await bundle.generate(outpuOptions)

  return result
}

export const getRollupOutput = async (absolutePathFileName) => {
  const inputOptions = {
    ...ROLLUP_INPUT_OPTIONS,
    input: absolutePathFileName,
  }
  const result = await runRollup(inputOptions, ROLLUP_OUTPUT_OPTIONS)
  return result.output[0].code.trim()
}

export const getRollupNoTreeshakingOutput = async (absolutePathFileName) => {
  const inputOptions = {
    ...ROLLUP_INPUT_OPTIONS,
    treeshake: false,
    input: absolutePathFileName,
  }
  const outpuOptions = {
    ...ROLLUP_OUTPUT_OPTIONS,
  }
  const result = await runRollup(inputOptions, outpuOptions)
  return result.output[0].code.trim()
}
