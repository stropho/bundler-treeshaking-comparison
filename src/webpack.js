import { createFsFromVolume, Volume } from 'memfs'
import webpack from 'webpack'
import TerserPlugin from 'terser-webpack-plugin'
import path from 'path'

const CONFIG = {
  mode: 'production',
  optimization: {
    minimize: true,
    minimizer: [
      new TerserPlugin({
        terserOptions: {
          mangle: false,
          output: { // default null,
            comments: 'all',
            beautify: true,
            quote_style: 1, // single quote
          },
        },
      }),
    ],
  },
  devtool: false,
  output: {
    path: __dirname,
    filename: 'output.js',
  },
}
export const runWebpack = (absolutePathFileName) => new Promise((resolve) => {
  const compiler = webpack({
    ...CONFIG,
    entry: absolutePathFileName,
  })
  const vol = new Volume()
  const fs = createFsFromVolume(vol)
  // for webpack@4
  fs.join = path.join.bind(path)
  compiler.outputFileSystem = fs

  // eslint-disable-next-line no-unused-vars
  compiler.run((err, stats) => {
    const output = fs.readFileSync(path.resolve(__dirname, './output.js'))
    resolve(output.toString())
  })
})

export const getWebpackExtractedCode = async (absolutePathFileName) => {
  const output = await runWebpack(absolutePathFileName)
  const code = output
    // remove webpackBootstrap before the actual code
    .replace(/[\s\S]*(function\([^{]*{[^\S]*["']use strict["'])/, '$1')
    // remove webpackBootstrap after the actual code
    .replace(/([\s\S]*)\/\*\*\*\*[\s\S]*/, '$1')

  return code.trim()
}
