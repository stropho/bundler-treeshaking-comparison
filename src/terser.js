import Terser from 'terser'
import { getRollupNoTreeshakingOutput } from './rollup'

export const TERSER_OPTIONS = {
  mangle: false,
  output: { // default null,
    comments: 'all',
    beautify: true,
    quote_style: 1, // single quote
  },
}

export const runTerserOnCombinedCode = async (absolutePathFileName) => {
  const options = {
    ...TERSER_OPTIONS,
    module: true,
  }
  const code = await getRollupNoTreeshakingOutput(absolutePathFileName)
  const result = Terser.minify(code, options)

  return result.code
}
