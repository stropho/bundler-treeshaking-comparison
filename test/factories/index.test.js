import path from 'path'
import { getRollupOutput, getWebpackExtractedCode, runTerserOnCombinedCode } from '../../src'

const FOLDER = path.basename(__dirname)
const INPUT_FILE = path.resolve(__dirname, './input.js')

describe(FOLDER, () => {
  test('rollup', async () => {
    const code = await getRollupOutput(INPUT_FILE)
    expect(code).toMatchSnapshot()
  })
  test('webpack + terser', async () => {
    const result = await getWebpackExtractedCode(INPUT_FILE)
    expect(result).toMatchSnapshot()
  })
  test('terser', async () => {
    const result = await runTerserOnCombinedCode(INPUT_FILE)
    expect(result).toMatchSnapshot()
  })
})
