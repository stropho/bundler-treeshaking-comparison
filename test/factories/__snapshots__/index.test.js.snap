// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factories rollup 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* eslint-disable operator-linebreak */
const reexport = 'reexport';

exports.reexport = reexport;"
`;

exports[`factories terser 1`] = `
"Object.defineProperty(exports, '__esModule', {
    value: !0
});

/* eslint-disable operator-linebreak */
const functionFactory = constantValue => () => constantValue;

functionFactory('unusedStringGetter');

exports.reexport = 'reexport';"
`;

exports[`factories webpack + terser 1`] = `
"function(module, __webpack_exports__, __webpack_require__) {
    'use strict';
    // ESM COMPAT FLAG
        __webpack_require__.r(__webpack_exports__), 
    // EXPORTS
    __webpack_require__.d(__webpack_exports__, 'reexport', (function() {
        /* reexport */ return reexport;
    }));
    // CONCATENATED MODULE: ./test/factories/exports.js
    /* eslint-disable operator-linebreak */
    const reexport = 'reexport', functionFactory = constantValue => () => constantValue;
    functionFactory('unusedStringGetter');
}"
`;
